File: core/src/main/java/net/particify/arsnova/core/validation/LanguageIso639.java
package net.particify.arsnova.core.validation;

import jakarta.validation.Constraint;
import jakarta.validation.Payload;
import java.lang.annotation.Documented;
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

@Constraint(validatedBy = LanguageIso639Validator.class)
@Documented
@Retention(RetentionPolicy.RUNTIME)
@Target({ElementType.FIELD, ElementType.METHOD, ElementType.PARAMETER})
public @interface LanguageIso639 {
  String message() default "Invalid ISO-639-1 language";
  Class<?>[] groups() default {};
  Class<? extends Payload>[] payload() default {};
}

File: core4/src/main/kotlin/net/particify/arsnova/core4/common/LanguageIso639.kt
/* Copyright 2025 Particify GmbH
 * SPDX-License-Identifier: MIT
 */
package net.particify.arsnova.core4.common

import jakarta.validation.Constraint
import jakarta.validation.Payload
import kotlin.reflect.KClass

@Constraint(validatedBy = [LanguageIso639Validator::class])
@MustBeDocumented
@Retention(AnnotationRetention.RUNTIME)
@Target(
    AnnotationTarget.FIELD,
    AnnotationTarget.FUNCTION,
    AnnotationTarget.PROPERTY_GETTER,
    AnnotationTarget.PROPERTY_SETTER,
    AnnotationTarget.VALUE_PARAMETER)
annotation class LanguageIso639(
    val message: String = "Invalid ISO-639-1 language",
    val groups: Array<KClass<*>> = [],
    val payload: Array<KClass<out Payload>> = []
)
